org.jacoco.core/src/org/jacoco/core/runtime/InjectedClassRuntime.java

```diff
@@ -0,0 +1,140 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2018 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.runtime;
+
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+/**
+ * {@link IRuntime} which defines a new class using
+ * {@code java.lang.invoke.MethodHandles.Lookup.defineClass} introduced in Java
+ * 9. Module where class will be defined must be opened to at least module of
+ * this class.
+ */
+public class InjectedClassRuntime extends AbstractRuntime {
+
+	private static final String FIELD_NAME = "data";
+
+	private static final String FIELD_TYPE = "Ljava/lang/Object;";
+
+	private final Class<?> locator;
+
+	private final String className;
+
+	/**
+	 * Creates a new runtime which will define a class to the same class loader
+	 * and in the same package and protection domain as given class.
+	 *
+	 * @param locator
+	 *            class to identify the target class loader and package
+	 * @param simpleClassName
+	 *            simple name of the class to be defined
+	 */
+	public InjectedClassRuntime(final Class<?> locator,
+			final String simpleClassName) {
+		this.locator = locator;
+		this.className = locator.getPackage().getName().replace('.', '/') + '/'
+				+ simpleClassName;
+	}
+
+	@Override
+	public void startup(final RuntimeData data) throws Exception {
+		super.startup(data);
+		Lookup //
+				.privateLookupIn(locator, Lookup.lookup()) //
+				.defineClass(createClass(className)) //
+				.getField(FIELD_NAME) //
+				.set(null, data);
+	}
+
+	public void shutdown() {
+		// nothing to do
+	}
+
+	public int generateDataAccessor(final long classid, final String classname,
+			final int probecount, final MethodVisitor mv) {
+		mv.visitFieldInsn(Opcodes.GETSTATIC, className, FIELD_NAME, FIELD_TYPE);
```

Comment:
`classname` vs. `className` is tricky. I would prefer a more distinct naming, e.g. rename member `className` to `injectedClassName`.

Reply:
@marchof done